tslint: use a better implementation of rule checking unused variables

bajtos opened the pull request:
The built-in `no-unused-variable` rule has many issues.
- It uses TSC API in a hacky way that breaks other rules depending on TSC API. See [tslint#2736](https://github.com/palantir/tslint/issues/2736) [tslint#2649](https://github.com/palantir/tslint/issues/2649) [tslint#2571](https://github.com/palantir/tslint/issues/2571)
- Because it depends on TSC API, it cannot be checked (and auto-fixed) from inside VS Code
- It is deprecated, `tslint` is printing the following warning now:
    ```
    no-unused-variable is deprecated. Since TypeScript 2.9.
    Please use the built-in compiler checks instead.
    ```
- It also cannot be used with [tslint as a compiler plugin](https://github.com/Microsoft/typescript-tslint-plugin).

I did an extensive search for alternatives and found `no-unused` from `tslint-consistent-codestyle` as the best alternative. See their [docs](https://github.com/ajafff/tslint-consistent-codestyle/blob/e5de338424cbb07acb727e9766e33a5d2eed367c/docs/no-unused.md) to learn more about the rule.
 
In this pull request:

- In the first commit, I am modifying our project templates to always install tslint and typescript dependencies, this is needed to satisfy requirements of `@loopback/tslint-config` and the new module `tslint-consistent-cdestyle`. 
- The second commit is removing `no-unused-variable` from `tslint.build.json` and configuring `no-unused` in `tslint.common.json`.
- The third commit is  fixing linting violations. This cannot be done as part of the previous commit because it would trigger semver-major release of all modified packages (at least I think so).

The third commit fixes configuration of VSCode tasks to use a correct problem matcher.

**BREAKING CHANGES**

This pull request introduces a breaking change in `@loopback/tslint-config`. People using comments to disable no-unused check for certain lines have to use a different directive after this change.

```diff
- // tslint:disable-next-line:no-unused-variable
+ // tslint:disable-next-line:no-unused
```

The new rule is also more strict and reports unused variables/parameters that are not detected by the built-in `no-unused-variable` rule.

cc @strongloop/loopback-next 

<!--
Please provide a high-level description of the changes made by your pull request.

Include references to all related GitHub issues and other pull requests, for example:

Fixes #123
Implements #254
See also #23
-->

## Checklist

- [x] `npm test` passes on your machine
- New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- API Documentation in code was updated
- Documentation in [/docs/site](../tree/master/docs/site) was updated
- Affected artifact templates in `packages/cli` were updated
- Affected example projects in `examples/*` were updated

bajtos commented:
For longer term, I'd like to refactor `@loopback/build` and extract TSLint configuration into a new package (e.g. `@loopback/tslint-config`) that can be versioned independently from the rest of our build tooling.

Maybe I should do that refactoring as the first step before this pull request, to avoid a breaking change in `@loopback/build`?

bajtos commented:
> Once tslint config has its own package, we can clearly communicate breaking changes in the autogenerated release notes.

See https://github.com/strongloop/loopback-next/pull/2159.

This PR is effectively blocked until #2159 is landed.

b-admike commented on packages/rest/test/acceptance/bootstrapping/rest.acceptance.ts:
```diff
@@ -76,7 +76,7 @@ async function startServerCheck(app: Application) {
 
 function sequenceHandler(
-  {request, response}: RequestContext,
+  {response}: RequestContext,
```
>On the other hand, the type RequestContext does contain both request & response, so I guess the "request" is still easy to discover even after this change.
I agree 👍

b-admike commented on packages/rest/test/acceptance/routing/routing.acceptance.ts:
```diff
@@ -692,7 +692,7 @@ describe('Routing', () => {
     it('provides httpHandler compatible with HTTP server API', async () => {
       const app = new RestApplication();
-      app.handler(({request, response}, sequence) => response.end('hello'));
+      app.handler(({response}, sequence) => response.end('hello'));
```
It's a bit harder to know that the `request` property is also available from the RequestContext than https://github.com/strongloop/loopback-next/pull/2156/commits/d14974af05311e61aa169bd1793d613b650f373e#r241689363.

b-admike commented on .vscode/tasks.json:
```diff
@@ -47,35 +47,7 @@
-            "message": 4
-          }
-        },
+        "$tslint5",
```
For my knowledge, can you clarify the benefits of using `tslint5` over the previous matcher algorithm?

bajtos replied:
Well the previous matcher was not working for me. I suspect it was based on `tslint4` format. I could not be bothered to figure out the right regular expressions and just picked `tslint5` because it seems to work well.

bajtos replied:
I decided to move this commit to a standalone PR to get it landed faster. See https://github.com/strongloop/loopback-next/pull/2162

bajtos commented:
Rebased on top of the new master. I need to fix the way how we are installing `tslint` to make linting work again, and update the commit message to clearly describe the breaking changes & migration guide for release notes.

bajtos commented on packages/metadata/src/types.ts:
```diff
@@ -17,6 +17,7 @@ export type DecoratorType =
  * @typeparam T Type of the metadata value
  * @typeparam D Type of the decorator
  */
+// tslint:disable-next-line:no-unused
```
This rule complains that there is no instance (prototype) method or property that would actually access T or D. Because these parameters are not used in such way, the rule things they should be removed.

On line 35, we are inside a static method and providing T and D values to create a specialized MetadataAccessor class. We are providing values to the two generic parameters, but these values won't be used anywhere inside the class.

Let me try to show this difference on a simpler example.

```ts
// This is similar to defining class MetadataAccessor<T, D>
function foo(unused: string) {
  // no-op
}

// This is similar to calling new MetadataAccessor<X, Y>(key)
foo('bar');
```

The function "foo" declares "unused" argument that's not accessed, therefore linter evaluates "unused" as something that can be removed.

Calling `foo('bar')` provides a value for the argument, but this does not make the argument "used", because the provided value is ignored by the function.

jannyHou replied:
Thank you for the detailed example, and

> This rule complains that there is no instance (prototype) method or property that would actually access T or D.

Makes sense 👍

bajtos commented:
In https://github.com/strongloop/loopback-next/pull/2159, I moved tslint config into a standalone package and changed our packages to access that config directly, bypassing `@loopback/build`. It turns out this approach is breaking tslint extensions which require access to tslint via a peer dependency.
```
project -> @loopback/tslint-config -> tslint-extension (requires tslint as a peer dep)
project -> @loopback/build -> tslint (tslint is in a different part of the dep tree)
```

I see two ways how to fix this, see below. Considering the pros and cons, I am leaning towards Option A.

@raymondfeng @strongloop/loopback-maintainers Which option do you prefer? Can you think of any better solution?

**Option A**

Continue slimming down `@loopback/build` and move both `tslint` and `typescript` from dependencies into peer-dependencies. (`tslint` has a peer dep on `typescript`, that's why both packages has to be moved together.) 

Pros: Users can decide what exact version of tsc & tslint they want to use and still keep using our builders. This can be very useful when tsc makes a breaking change in semver-minor release as it often does. 

Also, users can decide what version of LoopBack's tslint config they want to use, independently on `@loopback/build` version. 

Cons: We will have to duplicate the specification of `typescript` and `tslint` dev-dependencies in many places. At minimum in all example apps plus monorepo's root.

**Option B**

Keep the tslint config bundled inside `@loopback/build`. I think it still makes sense to have `@loopback/tslint-config` package (no changes there), but we need to revert the part of #2159 where we changed individual tslint config files (and the project template) to use `@loopback/tslint-config` directly.

Pros: No duplicated dev-dependencies, everything stays under `@loopback/build` umbrella.
Cons: Any breaking change in the tslint config will trigger a semver-major release of `@loopback/build` with all the implications for long term support (see [Making breaking changes](https://github.com/strongloop/loopback-next/blob/master/docs/site/DEVELOPING.md#making-breaking-changes)). Because most tslint config changes are backwards incompatible, we will be releasing semver-major versions of `@loopback/build` often.

Also, users of `@loopback/build` cannot decide which version of our tslint rules they want to use, they have to always use the version bundled inside the particular version of `@loopback/build` they depend on. (Same goes for tslint and tsc.)

bajtos commented:
Upon further investigation of Option A, it turns out that removing build's dependency on typescript and tslint is too complicated as it breaks other things (e.g. integration tests). I think it's enough to change our project templates to add typescript and tslint as explicit dev-dependencies, and fix build to prefer project-local version of typescript/tslint over the version installed as build's dependency - see https://github.com/strongloop/loopback-next/pull/2211

bajtos commented:
Travis CI is failing in code-lint step, I'll investigate the problem next week.

```
> node packages/build/bin/run-tslint --project tsconfig.json
Failed to load /home/travis/build/strongloop/loopback-next/tslint.build.json: 
  Could not find custom rule directory: tslint-consistent-codestyle
```

jannyHou commented on bin/sync-dev-deps.js:
```diff
@@ -0,0 +1,63 @@
+// License text available at https://opensource.org/licenses/MIT
+
+/**
+ * This is an internal script to synchronize versions of dev-dependencies
```
@bajtos When should we use this command? My guess is when updating a tslint/typescript version in the root level package.json?
A nitpick: maybe add some explanation in the documentation to remind developers to run it when updating dependencies.

bajtos commented:
> just one nitpick suggestion for adding some doc.
> When should we use this command? My guess is when updating a tslint/typescript version in the root level package.json?
> A nitpick: maybe add some explanation in the documentation to remind developers to run it when updating dependencies.

Good catch, I was thinking about the same too 👍
